### ISSUE TITLE
Convert Event-List block into query block-variation

### ISSUE THREAD

**Issue by carstingaxion** [GatherPress/gatherpress#599]
***tl/dr:*** 
Install the demo plugin using [gatherpress-query-loop.zip](https://github.com/carstingaxion/additional-advanced-query-loops/archive/refs/heads/gatherpress-query-loop.zip) AND checkout the corresponding [experimental branch of the GatherPress plugin](https://github.com/carstingaxion/gatherpress/tree/feature/experimental-query-block) to try out my exploration.

# Goal & Intent

*I wanted* to proof *my* idea, that it should be relatively simple to convert the existing GatherPress *Event List* block into a tailored block-variation of the core *Query* block.

Using a native, core ~ instead of a custom block has the following advantages:

1. Allows endless customisation in terms of layout & style
2. Reduce the number of strings to translate and comes with many translations for the core-strings.
3. Allows users to select from explicit compositions of block-patterns directly via the blocks **Choose (layout)** button and the **Replace** button in the top toolbar.

   https://github.com/figuren-theater/concept/assets/198883/749a54fa-d6fa-4642-b9c8-fe1d6f69ce86

4. Allows users to choose starter-content via the blocks **Start blank** button and the underlying [Block Variation Picker](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/block-variation-picker).

   https://github.com/figuren-theater/concept/assets/198883/5e7affb9-6ab6-49ec-906f-515f1a2283ac

5. Allows developers to register custom taxonomies with events, which would allow users to query by those taxonomies using native UI elements of the query block.

# Steps I took

- At first I started creating a block-variation for the query block from scratch  ... but realised very soon, that most of what's going to be needed, is already in the [**Advanced query loop** plugin](https://github.com/ryanwelcher/advanced-query-loop) by @ryanwelcher.
- So I forked it as a kind of [feature plugin for GatherPress](https://github.com/carstingaxion/additional-advanced-query-loops/tree/gatherpress-query-loop). At this point I was unsure about the amount of changes I would have to do, so it seemed cleaner to me, instead of experimenting within a huge plugin like GatherPress directly.
- I [renamed all slugs and prefixes](https://github.com/carstingaxion/additional-advanced-query-loops/commit/a8a52fd6e5861062c5a9b96f9527c36da452e48c) etc. from `advanced` or `aql` to `gatherpress` or `gpql`.
- Further on, I disabled some of the functionality, just by commenting out code.
- Disabled some of the native block-controls of the query-block, like the post type select.
- But kept some of the additional components from the advanced-query-loop plugin, that could be helpful or seemed interesting to us (or at least to me). 
- I [had to revert](https://github.com/carstingaxion/additional-advanced-query-loops/commit/e022113566b768fd41b9de3b34f717c4e22cdba4) https://github.com/GatherPress/gatherpress/issues/207, to be able to use all of the existing blocks *outside* of event- and venue-posts and within a query block.
- Got this working on the frontend, because of the nature of dynamic blocks and that they already rely on `get_the_ID()`. *Nice!*

   ![screenshot-rocks(2)](https://github.com/figuren-theater/concept/assets/198883/96bd4c5b-2bce-4648-a6b6-b3b909475b64)
   
   ![screenshot-rocks(3)](https://github.com/figuren-theater/concept/assets/198883/4ffafdf5-614f-4662-aed0-0ab9c34456c6)
   
   https://github.com/figuren-theater/concept/assets/198883/ef140815-d693-4225-a0f1-2e063055244c

- Until this point I was able to handle every change within the gatherpress-query-loop feature-plugin, but at some moment I (saw no way around, that I) needed to do some changes to GatherPress itself, so I created the earlier mentioned experimental-branch to also hold some of the changes.

- Tried to make the existing blocks aware of `usesContext: postId` to also get this working in the backend ... which was my weekends rabbit hole. 

   I mainly focused on the `gatherpress/event-date` and the `gatherpress/venue` blocks. And totally struggled, getting the blocks to work with the current `post ID` instead of the *event-list* logic.

   https://github.com/figuren-theater/concept/assets/198883/ac7a97e3-cc6d-486c-93a3-60bcdf29de8e

- I got this partly working for the *Venue* block, to show the `venue_information` in front- AND backend. Unfortunately `online-event`s are still only shown in the frontend, yet. 

---

**All of this is by far not finished, but I needed to get this out.**
**This all is still an absolute mess on its own, a mess in terms of performance and not DRY at all, but *it works and should help* illustrate the idea.**
*What do you think?*

# Next steps

- [x] Would it make sense to port this *feature-plugin* or parts of it into GatherPress core?

*If yes, ... we should ...*

- [ ] Make the *Event Date* Block aware of `usesContext: postId`
- [x] Make the *Venue* Block aware of `usesContext: postId`
- [x] Make the *Online-Event* Block aware of `usesContext: postId`
- [ ] Make the *Attending* Block aware of `usesContext: postId`

# Questions
- [x] What is or was the purpose of the `'gp_general''pages'` option, as it seems to become obsolete within block themes in general and with my suggested changes in particular?!
- [x] Which parts of the GatherPress code-base could help me while refactoring the existing blocks to utilise `usesContext`?
- [x] Could I use one of the existing REST API endpoints to get datetime and venue information together for a given event?
- [x] You (as a team) know the code and especially the inner workings of the blocks much better than me. Would you either recommend to modify or replace the existing blocks to make them work inside *query* blocks?
- [ ] What Kind of pattern would you like to see 
   a) for the **Choose Layout** modal
   b) for the **Start blank** option?
- [ ] Should I or we provide any kind of block-deprecation logic for the *Event List* block, or can I or we be lazy about this breaking-change as long as we don't have launched the magical v1.0?
- [ ] I would love to get a sortable, tabular view working! *Is anybody interested in exploring https://github.com/WordPress/gutenberg/issues/25503 together with me?* This could be one next big thing on the GatherPress journey.

**Comment by mauteri** [comment]
@carstingaxion I'm still reviewing all your notes, but I did download your work and tested it and was very cool. I think this could be a really cool change that could deprecate Events List with the release of WP 6.5 which leverages the query block and the interactivity API. One of the things I'd like to keep that is a feature of the Events List is that you can RSVP to an event from there, which is a dynamic feature of that block. I think we can replace a lot of the custom dynamic elements of our blocks and leverage the interactivity API, which is a better solution than the one we currently have (which is there because such things didn't exist when the blocks were initially built).

I'd also love to give you a more detailed tour of the plugin. Would you be able to jump on a Zoom with me for half hour or 45 minutes for me to take you through the plugin and some of the decisions that were made that we can discuss and think over. Let me know and thanks for doing this deep dive into our blocks!

**Comment by mauteri** [comment]
Docs on Interactivity API coming in WP 6.5 which I think we can leverage for many of our more dynamic blocks https://make.wordpress.org/core/2024/03/04/interactivity-api-dev-note/

**Comment by carstingaxion** [comment]
_NOTE:_ I updated some links in my initial comment, because I had to do some housekeeping & renamed the repository containing my experiment from 'gatherpress-query-loop' to '**additional-advanced-query-loops**', which now has a dedicated branch of `gatherpress-query-loop`.

**Comment by carstingaxion** [comment]
Started working on the `ORDER BY` of the query:

- [x] datetime (default)
- [x] random
- [ ] title
- [ ] post_id
- [ ] last modified date

**Comment by carstingaxion** [comment]
> [David F Carr](https://twitter.com/davidfcarr), developer of the [RSVP Maker plugin](https://wordpress.org/plugins/rsvpmaker/) and online publishing expert, shared his approach of [Transforming a Single WordPress Block into a Block Variation with InnerBlocks. ](https://davidfcarr.com/transforming-a-single-wordpress-block-into-a-block-variation-with-innerblocks/)“The virtue of block variations is they allow you to inherit all the best qualities of an existing block while adding a few tweaks, such as an InnerBlocks template or additional controls.” he wrote. The real-world examples and the shared code will “shortcut the process for someone else who might face a similar scenario.” Carr hopes.

https://gutenbergtimes.com/site-builders-showing-off-gutenberg-18-3-alternatives-to-meta-boxes-blueprint-gallery-weekend-edition-294/

**Comment by carstingaxion** [comment]
While working on the GatherPress query block I was annoyed by some core behaviour, so I also created a PR for that. And got it merged, [my first ever code contribution to Gutenberg](https://github.com/WordPress/gutenberg/pull/62277). Hooray!🎉

**Comment by MervinHernandez** [comment]
✅ Viewed 599 - Query Loop Block compliance - Carseten has it

**Comment by carstingaxion** [comment]
Finished `ORDER BY`

![image](https://github.com/GatherPress/gatherpress/assets/198883/67bb4e9f-0307-49fa-9a36-a074837ddd09)

**Comment by carstingaxion** [comment]
Implemented a new control, which allow for inclusion or exclusion of started, but not finished, events.


https://github.com/GatherPress/gatherpress/assets/198883/bb2ce6cc-5a5b-457d-9065-86c3f0a6aa40

**Comment by MervinHernandez** [comment]
✅ Reviewed June 22 - Carsten has it

**Comment by carstingaxion** [comment]
Fixed the UX fail, @patriciabt mentioned in slack, that's confussing to have two blocks to query events, the regular `core/query` block and the new `Event Query` block. _I agree._


https://github.com/user-attachments/assets/88f71ae8-d1e9-4d2d-ac12-ffff57e4d2bd

**Comment by mauteri** [comment]
> The related PR #962 is finished and ready to be reviewed and merged.
> 
> Please leave your feedback if you’re interested in a new event query block!

Let's have @patriciabt or someone test this as well from a user standpoint. Thanks!

**Comment by maninthemiddle01** [comment]
Hello!

First, thank you so much for your hard work on the new Events Query block variation. It's a fantastic feature for building flexible layouts with GatherPress!

I'm trying to create a list of only past events (like an event archive): when I select "Events" in the Query Loop, it correctly transforms into the "Events Query" block, but I can't seem to find a setting to filter for past vs. upcoming events.

I might be overlooking something, but is there a way to do this that I'm missing? I know the older "Events List" block has this toggle, and it would be amazing to have that same power in this new, more flexible query block, but somehow it's also a bit buggy, or I do something wrong, see #1392.

Since the goal of this issue was to create a powerful query variation, and filtering by date is such a common need for events, would it make sense to reopen this issue to track adding that feature? Or would you prefer I open a new, more specific issue for it?

Happy to provide more details or help test any solutions. Thanks again for all your great work on this plugin

**Comment by mauteri** [comment]
Hi @maninthemiddle01, yes you can use the `Event Query` block which can almost completely replace the `Event List` block. Currently, the `Online Event` and `Venue` blocks do not work in it, but will work once we release `0.34.0` and completely remove and replace the `Event List` block with the `Event Query` block, which will have the same look and feel, but be much more flexible and modern.

To set the `Event Query` block to Past vs Upcoming, select the `Event Query` block and look in the right panel under `Event Query Settings`. There you will see a toggle for `Upcoming or past events`. Toggling that will switch between the two. Let us know if you have any other questions. Thanks!
